src/EventTeachingService.cpp

```diff
@@ -0,0 +1,261 @@
+// Copyright (C) Sven Rosvall (sven@rosvall.ie)
+// This file is part of VLCB-Arduino project on https://github.com/SvenRosvall/VLCB-Arduino
+// Licensed under the Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License.
+// The full licence can be found at: http://creativecommons.org/licenses/by-nc-sa/4.0/
+
+#include <Streaming.h>
+
+#include "EventTeachingService.h"
+#include <Controller.h>
+#include <vlcbdefs.hpp>
+
+namespace VLCB {
+
+void EventTeachingService::setController(Controller *cntrl)
+{
+  this->controller = cntrl;
+  this->module_config = cntrl->module_config;
+}
+
+Processed EventTeachingService::handleMessage(unsigned int opc, CANFrame *msg)
+{
+  unsigned int nn = (msg->data[1] << 8) + msg->data[2];
+  unsigned int en = (msg->data[3] << 8) + msg->data[4];
+  // DEBUG_SERIAL << ">CbusSvc handling message op=" << _HEX(opc) << " nn=" << nn << " en" << en << endl;
+
+  switch (opc)
+  {
+    
+  case OPC_NNLRN:
+    // received NNLRN -- place into learn mode
+     DEBUG_SERIAL << F("> NNLRN for node = ") << nn << F(", learn mode on") << endl;
+
+    if (nn == module_config->nodeNum)
+    {
+      controller->bLearn = true;
```

Comment:
Sorry, can't add comment for this in Controller.h. Now that you have moved the code here, all uses of bLearn are in this service. Can you move this flag from Controller to the EventTeachingService?
I also see that bModeChanging and timeOutTimer in the controller are not used anymore since you added the MinimumNodeService. Can you remove them from the Controller?
(These are the kind of help I get from CLion, the IDE I talked about. ;) )

Reply:
Done